### ISSUE TITLE
`e_` cleanup: remove all passing of `evar_map`s by using references.

### ISSUE THREAD

**Issue by ejgallego** [coq/coq#6342]
Some parts of the codebase still manipulate `evar_map`s by using references, such as:

```ocaml
val pretype_type :
  int -> bool -> val_constraint -> env -> evar_map ref ->
  ltac_var_map -> glob_constr -> unsafe_type_judgment
```

We should indeed deprecate all the functions using this style and replace them with the equivalent functional version. It would be great to do the deprecation in 8.8.

A tree marking all the functions to be deprecated is at https://github.com/ejgallego/coq/tree/api%2Bevar_map_no_refs , so it could serve as a starting point. The number of warnings generated does indeed indicates that quite a bit of progress should be doable for 8.8.

A particular question is, what should we recommend for functions returning a pair, `Evd.sigma` ?

c.f. #946

**Comment by ejgallego** [comment]
What does remain to be done here?

**Comment by SkySkimmer** [comment]
Cases/pretyping are still using evdrefs I think.

**Comment by ejgallego** [comment]
This will be closed when #8580 #8581 #8582 and #8493 [among others] are merged.

Minor parts of internal code will still be using some evar passing by reference, but shouldn't be visible from outside.
